File: N-Queens II/N-Queens II.cpp
class Solution {
public:

    bool isValid(std::vector<int>& solution, int row, int col){
        for(int i = 0; i < solution.size(); ++i){
            if(solution[i] == col || abs(solution[i] - col) == abs(i - row)){
                return false;
            }
        }
        return true;
    }

    void recur(std::vector<int>& solution, int n, int level, int& counter){
        if(level == n){
            ++counter;
            return;
        }
        for(int i = 0; i < n; ++i){
            if(isValid(solution, level, i)){
                solution.push_back(i);
                recur(solution, n, level + 1, counter);
                solution.pop_back();
            }
        }
    }

    int totalNQueens(int n) {
        int counter = 0;
        std::vector<int> solution;
        recur(solution, n, 0, counter);
        return counter;
    }
};

File: N-Queens II/N-Queens II.java
public class Solution {
    
    private boolean isValid(int[] board, int row, int col){
        for(int i = 0; i < row; ++i){
            // If queens on two rows share the same column or diagonal
            // Return false
            if((col == board[i]) || (Math.abs(row - i) == Math.abs(col - board[i]))){
                return false;
            }
        }
        return true;
    }
    
    private void recur(int[] board, int row, int n, int[] counter){
        if(row == n){
            ++counter[0];
            return;
        }
        
        for(int i = 0; i < n; ++i){
            if(isValid(board, row, i)){
                board[row] = i;
                recur(board, row + 1, n, counter);
                board[row] = -1;
            }
        }
        
        return;
    }
    
    public int totalNQueens(int n) {
        int[] counter = new int[1];
        int[] board = new int[n];
        counter[0] = 0;
        for(int i = 0; i < n; ++i){
            board[i] = -1;
        }
        recur(board, 0, n, counter);
        return counter[0];
    }
}
